File: ios/AdjutantKit/Sources/AdjutantKit/Models/Kanban.swift
import Foundation
import SwiftUI

/// Kanban column identifiers matching bead status values.
/// Workflow: open -> hooked -> in_progress -> closed
public enum KanbanColumnId: String, Codable, CaseIterable, Identifiable {
    case open
    case hooked
    case inProgress = "in_progress"
    case closed

    public var id: String { rawValue }
}

/// Column configuration for the Kanban board.
public struct KanbanColumn: Identifiable {
    public let id: KanbanColumnId
    public let title: String
    public var beads: [BeadInfo]
    public let color: Color

    public init(id: KanbanColumnId, title: String, beads: [BeadInfo] = [], color: Color) {
        self.id = id
        self.title = title
        self.beads = beads
        self.color = color
    }
}

/// Column definition without beads, used for configuration.
public struct KanbanColumnDefinition {
    public let id: KanbanColumnId
    public let title: String
    public let color: Color

    public init(id: KanbanColumnId, title: String, color: Color) {
        self.id = id
        self.title = title
        self.color = color
    }
}

/// All column definitions in workflow order with Pip-Boy theme colors.
public let kanbanColumns: [KanbanColumnDefinition] = [
    KanbanColumnDefinition(id: .open, title: "OPEN", color: Color(hex: 0x00FF00)),
    KanbanColumnDefinition(id: .hooked, title: "HOOKED", color: Color(hex: 0x00FFFF)),
    KanbanColumnDefinition(id: .inProgress, title: "IN PROGRESS", color: Color(hex: 0x00FF88)),
    KanbanColumnDefinition(id: .closed, title: "CLOSED", color: Color(hex: 0x444444)),
]

/// Column definitions for Swarm mode (no HOOKED column).
/// Hooked beads are mapped to IN PROGRESS.
public let kanbanColumnsSwarm: [KanbanColumnDefinition] = [
    KanbanColumnDefinition(id: .open, title: "OPEN", color: Color(hex: 0x00FF00)),
    KanbanColumnDefinition(id: .inProgress, title: "IN PROGRESS", color: Color(hex: 0x00FF88)),
    KanbanColumnDefinition(id: .closed, title: "CLOSED", color: Color(hex: 0x444444)),
]

/// Returns the appropriate column definitions based on mode.
public func getKanbanColumns(isSwarm: Bool) -> [KanbanColumnDefinition] {
    return isSwarm ? kanbanColumnsSwarm : kanbanColumns
}

/// Maps bead statuses to Kanban columns.
/// In Swarm mode, hooked maps to inProgress.
public func mapStatusToColumn(_ status: String, isSwarm: Bool = false) -> KanbanColumnId {
    let normalized = status.lowercased()

    switch normalized {
    case "open":
        return .open
    case "hooked":
        return isSwarm ? .inProgress : .hooked
    case "in_progress":
        return .inProgress
    case "closed":
        return .closed
    default:
        return .open  // Unknown statuses default to open
    }
}

// MARK: - Color Extension

extension Color {
    /// Initialize Color from hex value
    init(hex: UInt) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0
        )
    }
}

File: frontend/src/types/kanban.ts
/**
 * Kanban workflow types for the Beads board.
 */

import type { BeadInfo } from './index';

/**
 * Valid Kanban column IDs matching bead status values.
 * Workflow: open -> hooked -> in_progress -> closed
 */
export type KanbanColumnId =
  | 'open'
  | 'hooked'
  | 'in_progress'
  | 'closed';

/**
 * Column configuration for the Kanban board.
 */
export interface KanbanColumn {
  id: KanbanColumnId;
  title: string;
  beads: BeadInfo[];
  color: string;
}

/**
 * All column definitions in workflow order (Gastown mode).
 */
export const KANBAN_COLUMNS: { id: KanbanColumnId; title: string; color: string }[] = [
  { id: 'open', title: 'OPEN', color: '#00FF00' },
  { id: 'hooked', title: 'HOOKED', color: '#00FFFF' },
  { id: 'in_progress', title: 'IN PROGRESS', color: '#00FF88' },
  { id: 'closed', title: 'CLOSED', color: '#444444' },
];

/**
 * Returns the Kanban column definitions.
 * Hooked beads are mapped to IN PROGRESS (no separate HOOKED column).
 */
export function getKanbanColumns(): { id: KanbanColumnId; title: string; color: string }[] {
  return [
    { id: 'open', title: 'OPEN', color: '#00FF00' },
    { id: 'in_progress', title: 'IN PROGRESS', color: '#00FF88' },
    { id: 'closed', title: 'CLOSED', color: '#444444' },
  ];
}

/**
 * Maps bead statuses to Kanban columns.
 * Hooked beads are mapped to in_progress.
 */
export function mapStatusToColumn(status: string): KanbanColumnId {
  const normalized = status.toLowerCase();

  switch (normalized) {
    case 'open':
      return 'open';
    case 'hooked':
    case 'in_progress':
      return 'in_progress';
    case 'closed':
      return 'closed';
    default:
      return 'open';
  }
}

/**
 * Drag state for Kanban interactions.
 */
export interface KanbanDragState {
  /** Bead being dragged */
  beadId: string | null;
  /** Source column */
  fromColumn: KanbanColumnId | null;
  /** Target column (during hover) */
  overColumn: KanbanColumnId | null;
}
